electrolytes/solvation_shell_extract.py

```diff
@@ -0,0 +1,266 @@
+
+    partial_charges = np.array(metadata["partial_charges"])
+
+    solutes = {
+        species: res
+        for res, species, type in zip(
+            metadata["residue"], metadata["species"], metadata["solute_or_solvent"]
+        )
+        if type == "solute"
+    }  # {species: residue name} mapping
+    solute_resnames = list(solutes.values())
+
+    solvents = {
+        species: res
+        for res, species, type in zip(
+            metadata["residue"], metadata["species"], metadata["solute_or_solvent"]
+        )
+        if type == "solvent"
+    }  # {species: residue name} mapping
+    solvent_resnames = list(solvents.values())
+
+    # Read a structure
+    structures = StructureReader(os.path.join(input_dir, "system_output.pdb"))
+
+    # For each solute: extract shells around the solute of some heuristic radii and bin by composition/graph hash
+    # Choose the N most diverse in each bin
+
+    for species, residue in solutes.items():
+        logging.info(f"Extracting solvation shells around {species}")
+        for radius in solute_radii:
+            logging.info(f"Radius = {radius} A")
+            expanded_shells = []
+            for i, st in tqdm(enumerate(structures)):  # loop over timesteps
+                # assign partial charges to atoms
+                for at, charge in zip(st.atom, partial_charges):
+                    at.partial_charge = charge
+
+                if i > 10:  # TODO: fix this
+                    break
+
+                # extract all solute molecules
+                solute_molecules = [
+                    res for res in st.residue if res.pdbres.strip() == f"{residue}"
+                ]
+                central_solute_nums = [mol.molecule_number for mol in solute_molecules]
+                # Extract solvation shells
+                shells = [
+                    set(
+                        analyze.evaluate_asl(
+                            st, f"fillres within {radius} mol {mol_num}"
+                        )
+                    )
+                    for mol_num in central_solute_nums
+                ]
+
+                # Now expand the shells
+                for shell, central_solute in zip(shells, central_solute_nums):
+                    expanded_shell = expand_shell(
+                        st,
+                        shell,
+                        central_solute,
+                        radius,
+                        solute_resnames,
+                        max_shell_size=200,
+                    )
+                    expanded_shells.append(expanded_shell)
+
+            # Now compare the expanded shells and group them by similarity
+            # we will get lists of lists of shells where each list of structures are conformers of each other
+            logging.info("Grouping solvation shells into conformers")
+            # TODO: speed this up
+            grouped_shells = group_with_comparison(expanded_shells, are_conformers)
```

Comment:
I'm not sure if we've discussed this before (probably we have at some point), but:

Right now, as written, we're only comparing structures _within a single MD run_. But we'll be doing multiple MD runs with (in some cases) very similar electrolytes. I'm not talking about different temperatures - I can see real benefit to separately sampling high-temp and low-temp structures. But what if we have KCl in water and NaCl in water? The water-water and water-Cl structures may well be very similar, and we may oversample certain interactions.

Reply:
Fair point, we should discuss this tomorrow. Maybe we'll need to have a final script that groups everything by conformers and then downselects. This will be expensive though, because `group_with_comparison` scales quadratically with the number of structures.

Reply:
We should discuss. Note that it doesn't quite scale quadratically in the number of structures, it goes as O(nm) where n is the number of structures and m is the number of distinct topologies. Assuming the number of different topologies is not strongly dependent on the number of systems, it's closer to linear scaling.